cl_dll/hud_strafeguide.cpp

```diff
@@ -0,0 +1,157 @@
+#include <complex>
+
+#include "hud.h"
+#include "cl_util.h"
+
+#include "pm_defs.h"
+#include "pm_movevars.h"
+
+enum border {
+	RED_GREEN,
+	GREEN_WHITE,
+	WHITE_GREEN,
+	GREEN_RED
+};
+
+const float pi = std::acos(-1.);
+
+int CHudStrafeGuide::Init()
+{
+	m_iFlags = HUD_ACTIVE;
+
+	hud_strafeguide = CVAR_CREATE("hud_strafeguide", "0", FCVAR_ARCHIVE);
+	hud_strafeguide_fov = CVAR_CREATE("hud_strafeguide_fov", "140", FCVAR_ARCHIVE);
+	hud_strafeguide_height = CVAR_CREATE("hud_strafeguide_height", "0", FCVAR_ARCHIVE);
+	hud_strafeguide_size = CVAR_CREATE("hud_strafeguide_size", "0", FCVAR_ARCHIVE);
+
+	gHUD.AddHudElem(this);
+	return 0;
+}
+
+int CHudStrafeGuide::VidInit()
+{
+	return 1;
+}
+
+int CHudStrafeGuide::Draw(float time)
+{
+	if (hud_strafeguide->value == 0)
+		return 0;
+	
+	double fov = hud_strafeguide_fov->value / 2 / 180 * pi;
+	
+	int size = gHUD.m_iFontHeight;
+	int height = ScreenHeight / 2 - 2*size;
+	
+	if (hud_strafeguide_size->value != 0)
+		size = hud_strafeguide_size->value;
+	
+	if (hud_strafeguide_height->value != 0)
+		height = hud_strafeguide_height->value;
+	
+	for (int i = 0; i < 4; ++i) {
+		int r, g, b;
+		switch (i) {
+			case RED_GREEN: case WHITE_GREEN:
+				r = 0; g = 255; b = 0; break;
+			case GREEN_WHITE:
+				r = 255; g = 255; b = 255; break;
+			case GREEN_RED:
+				r = 255; g = 0; b = 0; break;
+		}
+		
+		double boxLeft  = -angles[i];
+		double boxRight = -angles[(i+1)%4];
+		if (std::abs(boxLeft-boxRight) < 1e-10)
+			continue;
+		
+		if (boxLeft > fov) {
+			if (boxRight < -fov || boxRight > boxLeft)
+				continue;
+			boxLeft = -fov;
+		}
+		else if (boxRight < -fov) {
+			if (boxLeft < boxRight)
+				continue;
+			boxRight = fov;
+		}
+		
+		
+		if (boxLeft  < -fov) boxLeft  = -fov;
+		if (boxRight >  fov) boxRight = fov;
+		
+		int boxLeftI  = boxLeft / fov * ScreenWidth / 2;
+		int boxRightI = boxRight/ fov * ScreenWidth / 2;
+		boxLeftI  += ScreenWidth / 2;
+		boxRightI += ScreenWidth / 2;
+		
+		FillRGBA(boxLeftI, height, boxRightI-boxLeftI, size, r, g, b, 60);
+	}
+	
+	return 0;
+}
+
+static double angleReduce(double a)
+{
+	double tmp = std::fmod(a, 2*pi);
+	if (tmp < 0) tmp += 2*pi;
+	if (tmp > M_PI) tmp -= 2*pi;
+	return tmp;
+}
+
+void CHudStrafeGuide::Update(struct ref_params_s *pparams)
+{
+	double frameTime = pparams->frametime;
+	auto input = std::complex<double>(pparams->cmd->forwardmove, pparams->cmd->sidemove);
```

Comment:
A bit cursed to use complex for vectors but I guess it works

Reply:
There is `Vector2D` in the SDK with math operators, I'd use it instead
https://github.com/YaLTeR/OpenAG/blob/master/cl_dll/util_vector.h

Reply:
`Vector2D` was the first thing I used, but I couldn't make the  `#include`s work out right to actually use it properly. Tracing it through, I think `hud.h` (which indirectly brings in `util_vector.h`) needs to be included before `cl_util.h`,  but `cl_util.h` defines a macro for `DotProduct` which is incompatible with the Vector2D function of the same name...
I'm not really a C++ programmer so at that point I decided to go with the simpler solution of complex numbers. Also makes it easier to get the angle of a vector because you can just use a complex logarithm lol.